### DISCUSSION TITLE
Using vd.input and completer

### DISCUSSION THREAD

**Discussion by ECon87** [saulpw/visidata#1025] [Q&A] [answered]
Hi,

I am trying to create an auto-completer/suggestion tool for when I edit a cell, which pulls values from other entries on the same column (similar to what most spreadsheet editors have). There might a function that already achieves this in visidata but I couldn't find it.

I am trying to use the _vd.input()_ method, but I cannot to get the completer or updater to work. I guess I am not sure exactly what type the completer and/or updater should be.

```
from visidata import Sheet, Column, BaseSheet, vd
from prompt_toolkit import prompt
from prompt_toolkit.completion import WordCompleter
import readline


@Column.api
def get_values2(col, rows):
    values = [w for w in col.getValues(rows)]
    values = sorted(list(set(values)))
    values_completer = WordCompleter(values)
    choice = vd.input(prompt='Categories: ', completer=values_completer)
    return choice


Sheet.addCommand('1', 'get-values2',
                 'cursorCol.setValues([cursorRow], editCell(cursorVisibleColIndex, value=cursorCol.get_values2(rows)))',
                 'get suggested values')
```

I have also tried to use the _prompt_ method from _prompt\_toolkit_ but it returns a python object instead of a string.

Any hints would be greatly appreciated.

Cheers.
Cheers

**Comment by saulpw** [comment, marked as answer]
Hi @ECon87, the *completer* is a function which gets the string up until the cursor, and an index (integer) which increases/decreases with repeated Tab/Shift-Tab.  The completer function should return the string to replace it with (or empty to leave it alone).

The *updater* function is called every keystroke or timeout, to update the screen outside the input area.

**Reply by ECon87** [reply]
Hi @saulpw,

First of all thanks for all your great work, and thank you for taking the time to reply to my question.

I guess I should try a different method for getting the auto-compelter.

**Reply by saulpw** [reply]
@ECon87 It seems like you're pretty close, actually, and might be able to get it work with the existing completion functionality with not too much work (at least in the small).  Another option is to pass the `values` you have as the `history=` kwarg, and then use Up/Down arrow to cycle through them, which might work decently if there are only a few values.
